core/NameRef.h

```diff
@@ -70,9 +71,22 @@ class NameRef final : private DebugOnlyCheck<NameRefDebugCheck> {
 private:
     const Name &data(const GlobalState &gs) const;
 
+    // NameKind takes up this many bits in _id.
+    static constexpr u4 KIND_BITS = 2;
+    static constexpr u4 ID_BITS = 32 - KIND_BITS;
+    static constexpr u4 ID_MASK = (1 << ID_BITS) - 1;
+
+    u4 _id;
+
+    u4 unsafeTableIndex() const {
+        return _id & ID_MASK;
+    }
+
 public:
     friend GlobalState;
     friend Name;
+    // TODO(jvilk): Delurk this friend once name table is broken up.
+    friend GlobalSubstitution;
 
     NameRef() : _id(0){};
```

Comment:
It might be worth putting a comment on this constructor to document the fact that the value `0` now also implies that there's no kind tag present, which is why it's OK to use value as the default for all the NameRef kinds.

Reply:
Done